File: Source/Seasonal/Terrain/TerrainChanger.cs
using System;
using static WCSharp.Api.Common;

public static class TerrainChanger
{
    public static int[] Terrains { get; private set; }
    public static int[] SafezoneTerrain { get; private set; }
    public static string NormalCliff = "cXc1";
    public static string DirtCliff = "cXc2";
    public static int LastWolfTerrain = 0;
    public static int LastSafezoneTerrain = 0;

    /// <summary>The theme currently applied. Tracked so Apply() knows whose
    /// OnDeactivate hook to run before switching to a new theme.</summary>
    public static SeasonTheme CurrentTheme { get; private set; }

    public static void Initialize()
    {
        try
        {
            Terrains = new int[Gamemode.NumberOfRounds];
            SafezoneTerrain = new int[Gamemode.NumberOfRounds];
            Apply(SeasonThemeRegistry.None);
            if (Gamemode.CurrentGameMode != GameMode.Standard) return;
            Apply(SeasonalManager.CurrentTheme);
        }
        catch (Exception e)
        {
            Logger.Critical($"Error in TerrainChanger.Initialize {e.Message}");
            throw;
        }
    }

    /// <summary>
    /// Fills the per-round terrain arrays from a theme, repaints the map, and runs the
    /// theme's override hooks. Replaces the old NoSeasonTerrain()/ChristmasTerrain()/
    /// ActivateChristmasTerrain() trio - works for any theme, including future seasons,
    /// and for themes that override behavior instead of (or alongside) plain terrain,
    /// like Halloween's lightning-ringed safezones.
    /// </summary>
    public static void Apply(SeasonTheme theme)
    {
        if (CurrentTheme != null && CurrentTheme != theme)
        {
            CurrentTheme.OnDeactivate?.Invoke();
        }

        for (var i = 0; i < Gamemode.NumberOfRounds; i++)
        {
            Terrains[i] = theme.GetTerrainForRound(i);
            SafezoneTerrain[i] = theme.SafezoneTerrain;
        }

        CurrentTheme = theme;
        theme.OnActivate?.Invoke();
        SetTerrain();
    }

    /// <summary>Repaints the terrain for the current round and runs the active theme's
    /// OnRoundChange hook, if any. Call again after a round change.</summary>
    public static void SetTerrain()
    {
        SetWolfRegionTerrain();
        SetSafezoneTerrain();
        CurrentTheme?.OnRoundChange?.Invoke();
    }

    // Compatibility wrappers - remove once you've confirmed nothing outside
    // these files calls them directly.
    public static void NoSeason() => Apply(SeasonThemeRegistry.None);
    public static void ActivateChristmasTerrain() => Apply(SeasonThemeRegistry.Christmas);

    private static void SetWolfRegionTerrain()
    {
        var round = Globals.ROUND > 1 ? DifficultyConfig.GetVirtualRound(Globals.ROUND) - 1 : 0;
        ChangeMapTerrain(LastWolfTerrain, Terrains[round]);
        LastWolfTerrain = Terrains[round];
    }

    private static void SetSafezoneTerrain()
    {
        var round = Globals.ROUND > 1 ? DifficultyConfig.GetVirtualRound(Globals.ROUND) - 1 : 0;
        ChangeMapTerrain(LastSafezoneTerrain, SafezoneTerrain[round]);
        LastSafezoneTerrain = SafezoneTerrain[round];
    }

    public static void ChangeMapTerrain(int tileToChange, int newTerrain)
    {
        var mapRect = Globals.WORLD_BOUNDS;
        var minX = mapRect.MinX;
        var minY = mapRect.MinY;
        var maxX = mapRect.MaxX;
        var maxY = mapRect.MaxY;

        for (var x = minX; x <= maxX; x += 128)
        {
            for (var y = minY; y <= maxY; y += 128)
            {
                var type = GetTerrainType(x, y);

                if (type == tileToChange)
                    SetTerrainType(x, y, newTerrain, -1, 1, 1);
            }
        }
    }
}

File: src/Seasonal/Terrain/TerrainChanger.ts
import { Logger } from 'src/Events/Logger/Logger'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { HolidaySeasons, Seasons } from '../Seasons'

export class TerrainChanger {
    public static Terrains: number[] = []
    public static SafezoneTerrain: number[] = []
    public static NormalCliff = 'cXc1'
    public static DirtCliff = 'cXc2'
    public static LastWolfTerrain = 0
    public static LastSafezoneTerrain = 0

    public static Initialize = () => {
        try {
            TerrainChanger.NoSeason()
            if (CurrentGameMode.active !== GameMode.Standard) return
            TerrainChanger.ChristmasTerrain()
            TerrainChanger.SetTerrain()
        } catch (e) {
            Logger.Critical(`Error in TerrainChanger.Initialize: ${e}`)
            throw e
        }
    }

    /// <summary>
    /// Sets the terrain based on current round. Includes seasonal terrains.
    /// </summary>
    public static SetTerrain = () => {
        TerrainChanger.SetWolfRegionTerrain()
        TerrainChanger.SetSafezoneTerrain()
    }

    public static NoSeason = () => {
        TerrainChanger.NoSeasonTerrain()
        TerrainChanger.SetTerrain()
    }

    public static ActivateChristmasTerrain = () => {
        TerrainChanger.ChristmasTerrain()
        TerrainChanger.SetTerrain()
    }

    private static NoSeasonTerrain = () => {
        TerrainChanger.Terrains[0] = FourCC('Lgrd')
        TerrainChanger.Terrains[1] = FourCC('Ygsb')
        TerrainChanger.Terrains[2] = FourCC('Vgrs')
        TerrainChanger.Terrains[3] = FourCC('Xhdg')
        TerrainChanger.Terrains[4] = FourCC('Ywmb')

        for (let i = 0; i < Globals.NumberOfRounds; i++) {
            TerrainChanger.SafezoneTerrain[i] = FourCC('Xblm')
        }
        TerrainChanger.SetTerrain()
    }

    private static ChristmasTerrain = () => {
        if (Seasons.getCurrentSeason() !== HolidaySeasons.Christmas) return
        /*        SafezoneTerrain[0] = FourCC("Xblm");
                SafezoneTerrain[1] = FourCC("Ksmb");
                SafezoneTerrain[2] = FourCC("Drds");
                SafezoneTerrain[3] = FourCC("Kdkt");
                SafezoneTerrain[4] = FourCC("Oaby");*/

        for (let i = 0; i < Globals.NumberOfRounds; i++) {
            TerrainChanger.SafezoneTerrain[i] = FourCC('Ibsq') // Icecrown Glaicer (Black Squares)
        }
        for (let i = 0; i < Globals.NumberOfRounds; i++) {
            TerrainChanger.Terrains[i] = FourCC('Nrck')
        }
    }

    private static SetWolfRegionTerrain = () => {
        const round = Globals.ROUND > 1 ? Globals.ROUND - 1 : 0
        TerrainChanger.ChangeMapTerrain(TerrainChanger.LastWolfTerrain, TerrainChanger.Terrains[round])
        TerrainChanger.LastWolfTerrain = TerrainChanger.Terrains[round]
    }

    private static SetSafezoneTerrain = () => {
        const round = Globals.ROUND > 1 ? Globals.ROUND - 1 : 0
        TerrainChanger.ChangeMapTerrain(TerrainChanger.LastSafezoneTerrain, TerrainChanger.SafezoneTerrain[round])
        TerrainChanger.LastSafezoneTerrain = TerrainChanger.SafezoneTerrain[round]
    }

    public static ChangeMapTerrain = (tileToChange: number, newTerrain: number) => {
        const mapRect = Globals.WORLD_BOUNDS
        const minX = mapRect.minX
        const minY = mapRect.minY
        const maxX = mapRect.maxX
        const maxY = mapRect.maxY

        for (let x = minX; x <= maxX; x += 128) {
            for (let y = minY; y <= maxY; y += 128) {
                const type = GetTerrainType(x, y)

                if (type === tileToChange) SetTerrainType(x, y, newTerrain, -1, 1, 1)
            }
        }
    }
}
